### PR TITLE
move prod/sum common code to composite.py; also add iter/len/getitem

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Prod and Sum include a lot of similar code. To follow DRY (don't-repeat-yourself) practice, I'm moving it into one file (composite.py) before making any more changes to either class. Then I'm making some changes :)

**Description of the Change:**
- Move prod/sum common code to composite.py
- Add `len`, `iter` and `getitem` dunder methods to the new CompositeOp class to help iterate over Sum and Prod operands

**Benefits:**
- We can iterate over composite operators now! (and other access methods like indexing and getting length)
- There is less copied code in the repo, making it easier to understand
- New changes that would have to be made to both prod and sum can now be applied in one place
- A potential new composite operator class would not need to re-implement anything
- Some extra safeguards (errors raised) if the child class is missing some abstract method implementations

**Possible Drawbacks:**
- If the two (or more) types diverge greatly, this abstraction can bring unneeded overhead. I don't think this is the case. They're fairly different as it is today, but there's enough overlap that this brings value. We can always add/remove to/from the base class if necessary.

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/ops/op_math/composite.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        if self._name is None:
+            raise NotImplementedError("Child class must specify _name")
+        if self._op_symbol is None:
+            raise NotImplementedError("Child class must specify _op_symbol")
+        if len(operands) < 2:
+            raise ValueError(f"Require at least two operators to combine; got {len(operands)}")
+
+        self.operands = operands
+        self._wires = qml.wires.Wires.all_wires([op.wires for op in self.operands])
+        self._hash = None
```
Comment (timmysilv) [inline pennylane/ops/op_math/composite.py:60]: I added this here to avoid Prod requiring an init dunder just to say `self._hash = None; super().__init__(*args)`

File: pennylane/ops/op_math/composite.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    @property
+    def num_params(self):
+        return sum(op.num_params for op in self.operands)
+
+    @property
+    @abc.abstractmethod
+    def is_hermitian(self):
+        """
+        TODO: should we do `all(op.is_hermitian for op in self.operands)` as default?
```
Comment (timmysilv) [inline pennylane/ops/op_math/composite.py:121]: open question

File: pennylane/ops/op_math/composite.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        """
+
+    @abc.abstractmethod
+    def matrix(self, wire_order=None):
+        """Representation of the operator as a matrix in the computational basis."""
+
+    @abc.abstractmethod
+    def sparse_matrix(self, wire_order=None):
+        """Compute the sparse matrix representation of the composite operator."""
```
Comment (timmysilv) [inline pennylane/ops/op_math/composite.py:130]: should we require all child classes implement this? the `SparseMatrixUndefinedError` will [come from the parent Operator class](https://github.com/PennyLaneAI/pennylane/blob/3306eddbf2822c4de958d75519d2e4a235c54dc5/pennylane/operation.py#L498-L539) anyway

File: tests/ops/op_math/test_composite.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    def test_diagonalizing_gates(self):
+        """Test that the diagonalizing gates are correct."""
+        diag_op = ValidOp(*self.simple_operands)
+        diagonalizing_gates = diag_op.diagonalizing_gates()
+
+        assert len(diagonalizing_gates) == 1
+        diagonalizing_mat = diagonalizing_gates[0].matrix()
+
+        true_mat = qnp.array(
+            [[1.0, 0.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 1.0, 0.0], [0.0, 0.0, 0.0, 1.0]]
```
Comment (timmysilv) [inline tests/ops/op_math/test_composite.py:154]: note this test is not super meaningful because the `ValidOp.matrix()` method just returns identity regardless of operands.. just showing it does some stuff :p maybe a more interesting forced matrix would be useful? The Prod and Sum unit tests still cover this.

File: tests/ops/op_math/test_composite.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    def test_eigen_caching(self):
+        """Test that the eigendecomposition is stored in cache."""
+        diag_op = ValidOp(*self.simple_operands)
+        eig_decomp = diag_op.eigendecomposition
+
+        eig_vecs = eig_decomp["eigvec"]
+        eig_vals = eig_decomp["eigval"]
+
+        eigs_cache = diag_op._eigs[
+            (1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0)
```
Comment (timmysilv) [inline tests/ops/op_math/test_composite.py:168]: same comment as for `test_diagonalizing_gates`

File: tests/ops/op_math/test_prod.py [hunk 6]
```diff
@@ -733,6 +633,11 @@ def test_depth_property(self):
         )
         assert prod_op.arithmetic_depth == 3
 
+        op_constructed = Prod(
+            qml.RZ(1.32, wires=0), qml.Identity(wires=0), qml.RX(1.9, wires=1), qml.PauliX(0)
+        )
+        assert op_constructed.arithmetic_depth == 1
```
Comment (timmysilv) [inline tests/ops/op_math/test_prod.py:639]: unfortunate discrepancy. same as in test_sum.py :/

File: pennylane/ops/op_math/composite.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        )
+
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.operands = tuple(s.__copy__() for s in self.operands)
+
+        for attr, value in vars(self).items():
+            if attr not in {"operands"}:
+                setattr(copied_op, attr, value)  # TODO: exclude data?
```
Comment (timmysilv) [inline pennylane/ops/op_math/composite.py:77]: open question - prod and sum used to exclude data, summands and factors fyi

**Review by timmysilv — COMMENTED** [review]
just a few things I want to point out to reviewers

File: pennylane/ops/op_math/composite.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        id (str or None): id for the operator. Default is None.
+
+    The child composite operator should define the `_name` and `_op_symbol` properties
+    during initialization and define any relevant representations, such as
+    :meth:`~.operation.Operator.matrix` and :meth:`~.operation.Operator.decomposition`.
+    """
+
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+    _op_symbol = None
+    _name = None
```
Comment (albi3ro) [inline pennylane/ops/op_math/composite.py:43]: Maybe instead of raising errors in `__init__`, we can make these abstract properties:
```
    @property
    @abc.abstractmethod
    def _op_symbol(self):
        pass
```

We might also be able to leave `_name` as just `self.__class__.__name__`, as is set in `Operator`.

File: pennylane/ops/op_math/composite.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        return self.operands[idx]
+
+    def __len__(self):
+        """Return the number of operators in this composite operator"""
+        return len(self.operands)
+
+    @property
+    def op_symbol(self) -> str:
+        """The symbol used when visualizing the composite operator"""
+        return self._op_symbol
```
Comment (albi3ro) [inline pennylane/ops/op_math/composite.py:96]: Does this need to be a public property?

File: pennylane/ops/op_math/composite.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/ops/op_math/composite.py:130]: I think we can leave these representations out.  If they are not implemented, then we should just have the base `Operator` behavior.

**Review by albi3ro — COMMENTED** [review]
I'm realizing now that `Sum` and `Prod` don't actually define `has_matrix`.  Not sure how this managed to slip our attention this whole time 😨 

Mind adding this property?
```
@property
def has_matrix(self):
    return all(op.has_matrix for op in self)
```

File: pennylane/ops/op_math/composite.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+    @property
+    def data(self):
+        """Create data property"""
+        return [op.parameters for op in self.operands]
```
Comment (albi3ro) [inline pennylane/ops/op_math/composite.py:101]:
```suggestion
        return [op.parameters for op in self]
```

Because we can do that now!

File: pennylane/ops/op_math/composite.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+    @data.setter
+    def data(self, new_data):
+        """Set the data property"""
+        for new_entry, op in zip(new_data, self.operands):
```
Comment (albi3ro) [inline pennylane/ops/op_math/composite.py:106]:
```suggestion
        for new_entry, op in zip(new_data, self):
```
Bunch of places where we can now just iterate over `self` 😄

File: pennylane/ops/op_math/composite.py (hunk 8 above)
Comment (timmysilv) [reply pennylane/ops/op_math/composite.py]: makes sense to me! updated

File: pennylane/ops/op_math/composite.py (hunk 9 above)
Comment (timmysilv) [reply pennylane/ops/op_math/composite.py]: nope. updated

File: pennylane/ops/op_math/composite.py (hunk 1 above)
Comment (timmysilv) [reply pennylane/ops/op_math/composite.py:55]: I see that a future change will incorporate this into Sum as well, no worries here

File: pennylane/ops/op_math/composite.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/ops/op_math/composite.py]: Prod is different enough from Sum and I have since made this abstract. Removed the todo

File: pennylane/ops/op_math/composite.py (hunk 10 above)
Comment (Jaybsoni) [reply pennylane/ops/op_math/composite.py]: I vaguely remember having some issues with aliasing in the data. When copying an operator, if the parameters didn't deep copy the data, then shallowing copying and manipulating the parameters changed the base objects data. 

I am pretty sure I added tests for these cases, so I think if the tests pass we should be okay 👍🏼

File: pennylane/ops/op_math/composite.py (hunk 7 above)
Comment (albi3ro) [reply pennylane/ops/op_math/composite.py:72]: Sum and Prod have no data of their own.  Copying their components essentially copies their data.

File: pennylane/ops/op_math/prod.py [hunk 12]
```diff
@@ -151,67 +151,21 @@ def circuit(weights):
...
 
-        for attr, value in vars(self).items():
-            if attr not in {"factors"}:
-                setattr(copied_op, attr, value)
-
-        return copied_op
+    @property
+    def factors(self):
+        """Return the factors that compose this operator."""
+        return self.operands
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:160]: I think we could actually just deprecate `factors` and `summands` in favour of the more general `operands`.

File: pennylane/ops/op_math/prod.py [hunk 13]
```diff
@@ -221,20 +169,10 @@ def is_hermitian(self):
         yields false, which ARE hermitian. So a false result only implies a more explicit check
         must be performed.
         """
-        for o1, o2 in combinations(self.factors, r=2):
+        for o1, o2 in combinations(self.operands, r=2):
```
Comment (albi3ro) [inline pennylane/ops/op_math/prod.py:172]: We could probably do this in another PR, but this function could now be replaced with:
```suggestion
        return False if self.has_overlapping_wires else all(op.is_hermitian for op in self)
```

Comment (timmysilv) [reply pennylane/ops/op_math/prod.py:172]: nice! I'll do it in another PR just in the name of time and simplicity but it's definitely better

File: tests/ops/op_math/test_prod.py (hunk 6 above)
Comment (Jaybsoni) [reply tests/ops/op_math/test_prod.py:621]: Ahh this was due to the lack of simplify ?

**Review by Jaybsoni — APPROVED** [review]
Amazing 💯 ! Everything was very straightforward! No questions / comments from my side!

File: tests/ops/op_math/test_prod.py (hunk 6 above)
Comment (timmysilv) [reply tests/ops/op_math/test_prod.py:621]: well, you could simplify the one before and it would also have a depth of 1 :) just no one did it in this test
